crates/downloader/src/download_queue.rs

```diff
@@ -0,0 +1,273 @@
+    async fn get_block(&self, block_number: u64) -> Result<Option<Block>, GetBlockError> {
+        self.get_block_by_number(BlockNumberOrTag::Number(block_number), true)
+            .await
+            .map_err(|err| match err {
+                RpcError::ErrorResp(resp) if resp.code == 429 => {
+                    GetBlockError::RateLimited
+                }
+                other => GetBlockError::Other(other),
+            })
+    }
+}
+
+enum Mode {
+    /// Batch-download blocks until we reach the end of the chain
+    Batch,
+    /// Poll for new blocks
+    Poll,
+}
+impl Mode {
+    /// How many blocks to download concurrently
+    fn concurrency(&self) -> usize {
+        match self {
+            Mode::Batch => 8,
+            Mode::Poll => 1,
+        }
+    }
+}
+
+pub struct DownloadQueue<P> {
+    provider: P,
+    next_da_height: u64,
+    pending: JoinSet<Result<Option<Block>, GetBlockError>>,
+    completed: BTreeMap<u64, Block>,
+}
+
+impl<P: GetBlock + Clone + Send + 'static> DownloadQueue<P> {
+    pub fn start_from(provider: P, da_height: u64) -> Self {
+        Self {
+            provider,
+            next_da_height: da_height,
+            pending: JoinSet::new(),
+            completed: BTreeMap::new(),
+        }
+    }
+
+    async fn stream_inner(
+        mut self,
+        tx: mpsc::Sender<Result<Block, anyhow::Error>>,
+    ) -> anyhow::Result<()> {
+        let mut next_to_emit = self.next_da_height;
+        let mut mode = Mode::Batch;
+
+        loop {
+            // Queue new downloads up to the concurrency limit
+            while self.pending.len() + self.completed.len() < mode.concurrency() {
+                let provider = self.provider.clone();
+                self.pending
+                    .spawn(async move { provider.get_block(self.next_da_height).await });
+                self.next_da_height += 1;
+            }
+
+            // Wait for the next block to complete
+            if let Some(block) = self.pending.join_next().await {
+                match block.expect("Failed to join block download task") {
+                    Ok(Some(block)) => {
+                        self.completed.insert(block.header.number, block);
+                    }
+                    Ok(None) => {
+                        if matches!(mode, Mode::Batch) {
+                            tracing::info!(
+                                "Latest block reached, moving to polling mode"
+                            );
+                            mode = Mode::Poll;
+                        } else {
+                            self.next_da_height = next_to_emit;
+                            if self.pending.is_empty() {
+                                tracing::trace!(
+                                    "Block is not yet available, trying again later."
+                                );
+                                tokio::time::sleep(std::time::Duration::from_secs(12))
```

Comment:
```suggestion
const DA_BLOCK_TIME_MILLIS: u32 = 12_000;                                tokio::time::sleep(std::time::Duration::from_millis(12_000))
```

or make it configurable to make testing a bit easier?

Reply:
but i see in the tests we're advancing time manually, so that works too :) the point is not to use magic numbers then

Reply:
addressed in 5b422d9